packages/go/analysis/post.go

```diff
@@ -120,6 +123,57 @@ type DeleteRelationshipJob struct {
 	ID   graph.ID
 }
 
+// EdgeConstraintMap is a thread safe tracker for post-processed edges. It guarantees that only one edge of a given
+// post-processed kind may exist. This is useful either to create a batch of post-processed edges to insert or to
+// guard against double insertion of the same post-processed edge.
+type EdgeConstraintMap struct {
+	lock     *sync.Mutex
+	adjacent map[graph.ID]map[graph.Kind]cardinality.Duplex[uint64]
+}
+
+func NewEdgeConstraintMap() EdgeConstraintMap {
+	return EdgeConstraintMap{
+		lock:     &sync.Mutex{},
+		adjacent: map[graph.ID]map[graph.Kind]cardinality.Duplex[uint64]{},
+	}
+}
+
+// TrackCreatePostRelationshipJob decomposes a CreatePostRelationshipJob type and returns the result of tracking it.
+func (s EdgeConstraintMap) TrackCreatePostRelationshipJob(job CreatePostRelationshipJob) bool {
+	return s.Track(job.FromID, job.ToID, job.Kind)
+}
+
+// Track will attempt to track creation of the given relationship arguments. This function returns false
+// if the given relationship has already been tracked; true otherwise.
+func (s EdgeConstraintMap) Track(start, end graph.ID, edgeKind graph.Kind) bool {
```

Comment:
Nit, this isn't used outside this file, can we swap this to a private function?

Reply:
I like this being a public contract for right now. It's the basic interaction of the datastructure and makes it more usable for little cost.